src/shared/telemetry/spans.ts

```diff
@@ -146,11 +147,18 @@ export class TelemetrySpan {
             duration,
             result: getTelemetryResult(err),
             reason: getTelemetryReason(err),
-        })
+        } as Partial<T>)
 
         this.#startTime = undefined
     }
 
+    public increment(data: { [P in NumericKeys<T>]+?: number }): void {
+        for (const [k, v] of entries(data)) {
+            ;(this.state as Record<typeof k, number>)[k] = ((this.state[k] as number) ?? 0) + v!
+        }
+    }
+
+    // TODO: implement copy-on-write abstraction if this method causes perf issues
```

Comment:
At that point we should look at a "persistent datatype" library rather than implementing this ourselves.

Reply:
Yeah this comment isn't relevant anymore either. In one commit I had `clone()` being called a lot more.